Config cleanup

MarkusG opened the pull request:
This addresses #130

* Remove the `DataBlock` struct and handle config files as a set of key-value pairs.
* Remove `LOCAL_HOST` constant and instead use `"127.0.0.1"` verbatim.
* `client.obd` and `server.obd` are now separate files.
* `loadFromConfigFile()` will now generate config files with the default values if none are found.

I tried to keep this PR as non-breaking as possible. However, I'd like to get your thoughts on reworking the configuration system to leverage OOP. Consider the following class:

```cpp
class ObdConfig
{
        private:
		std::unordered_map<std::string, std::string> m_keyvalues;
	public:
		ObdConfig(const std::string& filePath); // implicitly read config file
		std::string get_option(const std::string& key);

		~ObdConfig(); // write back to file when the config goes out of scope
};
```

This way, we can centralize all the config logic into its own class and keep main clean. Also, by utilizing constructors and destructors, we can keep even more code out of main. The only changes that would need to be made to the rest of the project are replacing references to `config.client.OPTION` with calls to  `clientConfig.get_option("OPTION")`. Let me know what you think of this alternative implementation and I'll add it on to this PR.

MarkusG committed:
Rework config files

Summary:
Simplify parsing of configuration files, reading them into a plain old
unordered_map and splitting server and client into two config files.

obd_parser:
* Remove DataBlock struct and helper functions in obd_parser.cpp and
  obd_parser.h
* Rename readObdData to parseObdData and simplify it to only take a
  string and return an unordered_map of key/value pairs

obd_parser_test:
* simplify test to ensure integrity of 3 values, including excess
  whitespace and newlines in string input

main:
* read from two files, client.obd and server.obd, rather than a single
  configuration file with CLIENT and SERVER sections
* include file_io header to read the contents of the config files as
  strings, and then pass the strings to parseObdData

MarkusG committed:
Generate default config

Summary:
Check if config files exist. If a given config file does not exist,
write the default values to a new config file.

Also, remove LOCAL_HOST constant as 127.0.0.1 is a standard for the
local host.

file_io:
Add new fileExists() function to check if a file exists

net_constants:
Remove LOCAL_HOST constant

net_host_tests:
Replace LOCAL_HOST with "127.0.0.1"

obd_parser:
Add serializeObdData() function to serialize a key-value collection to
the obd format

main:
Rewrite loadFromConfigFile() function to handle a non-present config
file.

MarkusG committed:
gitignore: ignore *.obd

b0guslaw reviewed (changes requested):
Disagree with excluding .obd files from git as new releases on master may require new config.obd files

Hopson97 commented:
Shouldn't be an issue, new config files would be generated if they do not exist (Which is one of the features of this PR)

I'll take a closer look/review this in a bit, thanks for this :)

b0guslaw commented:
Complete oversight on my part, disregard my Request for change

Hopson97 commented on src/common/common/file_io.cpp:
```diff
@@ -15,3 +15,9 @@ std::string loadFileContents(const std::string& path)
+{
+    std::ifstream inFile(path);
+	return !!inFile;
+}
```
C++17 has `<filesystem>`, which allows you to do `std::filesystem::exists(path)`

https://en.cppreference.com/w/cpp/filesystem/exists

So this function isn't necessarily needed (or it can be simple wrapper around the c++17 function, which looks kinda long)

Hopson97 commented on src/common/common/obd_parser.cpp:
```diff
@@ -6,51 +6,26 @@
-    std::string line;
+	while (contents >> key) {
+		contents >> value;
+		data.insert(std::make_pair(key, value));
```
`std::unordered_map::emplace` can be used here

https://en.cppreference.com/w/cpp/container/unordered_map/emplace

This would shortern this code to `data.emplace(key, value);` as the pair would be constructed in place

Hopson97 commented on src/main.cpp:
```diff
@@ -49,27 +50,59 @@ struct Config {
+		clientData.insert(std::make_pair("fps_limit", std::to_string(config.client.fpsLimit)));
+		clientData.insert(std::make_pair("skin", config.client.skinName));
+		clientData.insert(std::make_pair("texture_pack", config.client.texturePack));
+		clientData.insert(std::make_pair("server_ip", config.client.serverIp));
```
Same thing here about https://en.cppreference.com/w/cpp/container/unordered_map/emplace

Hopson97 commented on src/main.cpp:
```diff
@@ -49,27 +50,59 @@ struct Config {
+	}
+	else {
+		std::unordered_map<std::string, std::string> serverData;
+		serverData.insert(std::make_pair("world_size", std::to_string(config.server.worldSize)));
```
https://en.cppreference.com/w/cpp/container/unordered_map/emplace again 😛

Hopson97 reviewed (changes requested):
Looking really good from a functionality point of view, a few suggestions to modernize the code a bit however :)

MarkusG committed:
Hopson97 code review

Summary:
Modernize code by using C++17 features and unordered_map::emplace() rather
than unordered_map::insert(std::make_pair(...))

file_io:
* Remove fileExists() function

obd_parser:
* Replace calls to unordered_map::insert() with calls to

main:
* Replace calls to unordered_map::insert() with calls to
unorderd_map::emplace()
* Use std::filesystem::exists() instead of fileExists helper function

MarkusG commented:
Revised according to review

Hopson97 reviewed (approved):
Perfect :)

MarkusG commented:
So what are your thoughts on the config class I mentioned earlier? I think it would be better to have the repetitive code that reads config files stowed away in its own class, and I like the idea of taking advantage of constructors and destructors to automatically read/write the files.

Hopson97 commented:
Yeah could make sense actually :)

Only issue is it feels weird to use a contructor and destructor this way, but I guess a simple comment could make it clearer for people reading the code.

I'd say go for it! :)

MarkusG committed:
Config overhaul

Summary:
Introduce three new classes: ObdConfig, ClientConfig, and ServerConfig.
ObdConfig provides a protected field of key-value pairs that are read in
from the configuration file on object construction and written on object
destruction. ClientConfig and ServerConfig derive from ObdConfig and
map the key-value pairs to their fields in their constructors and
destructors. New config classes use private fields and get_value() and
set_value() accessors in accordance with C++ best practices. These three
new classes replace the previously used Config, ClientConfig, and
ServerConfig structs.

All files:
* Update references to old config files (e.g. config.client.field) for
  new config classes (e.g. client.get_field())
* Replace references to old Config struct with references to both the
  ClientConfig and ServerConfig objects

obd_parser:
* Move functions previously located in obd_parser to obd_config

main:
* Remove Config struct
* Remove loadFromConfigFile() function
* Simplify parseArgs() function to only handle parsing options,
  leaving validation elsewhere in main()
* Reference ClientConfig, ServerConfig, and LaunchType individually
  instead of the old Config struct

MarkusG commented:
Totally overhauled the config system. Here's 79a40d's commit message that summarizes the changes:
```
    Config overhaul
    
    Summary:
    Introduce three new classes: ObdConfig, ClientConfig, and ServerConfig.
    ObdConfig provides a protected field of key-value pairs that are read in
    from the configuration file on object construction and written on object
    destruction. ClientConfig and ServerConfig derive from ObdConfig and
    map the key-value pairs to their fields in their constructors and
    destructors. New config classes use private fields and get_value() and
    set_value() accessors in accordance with C++ best practices. These three
    new classes replace the previously used Config, ClientConfig, and
    ServerConfig structs.
    
    All files:
    * Update references to old config files (e.g. config.client.field) for
      new config classes (e.g. client.get_field())
    * Replace references to old Config struct with references to both the
      ClientConfig and ServerConfig objects
    
    obd_parser:
    * Move functions previously located in obd_parser to obd_config
    
    main:
    * Remove Config struct
    * Remove loadFromConfigFile() function
    * Simplify parseArgs() function to only handle parsing options,
      leaving validation elsewhere in main()
    * Reference ClientConfig, ServerConfig, and LaunchType individually
      instead of the old Config struct
```

MarkusG committed:
Revert "Config overhaul"

Did I break something or did the unit tests break???

MarkusG committed:
Revert "Revert "Config overhaul""

Yep I broke something.

MarkusG commented:
Now I'm confused. It looks like the CI build is failing because it can't find the `filesystem` header, but faca4153d816b4067fe64822c47880737a33361d references it and it passed the CI build just fine.
CI build output:
```
Scanning dependencies of target ob-common
[  1%] Building CXX object src/common/common/CMakeFiles/ob-common.dir/file_io.cpp.o
[  2%] Building CXX object src/common/common/CMakeFiles/ob-common.dir/obd_config.cpp.o
/home/runner/work/open-builder/open-builder/src/common/common/obd_config.cpp:5:10: fatal error: filesystem: No such file or directory
 #include <filesystem>
          ^~~~~~~~~~~~
compilation terminated.
src/common/common/CMakeFiles/ob-common.dir/build.make:75: recipe for target 'src/common/common/CMakeFiles/ob-common.dir/obd_config.cpp.o' failed
CMakeFiles/Makefile2:327: recipe for target 'src/common/common/CMakeFiles/ob-common.dir/all' failed
make[2]: *** [src/common/common/CMakeFiles/ob-common.dir/obd_config.cpp.o] Error 1
make[1]: *** [src/common/common/CMakeFiles/ob-common.dir/all] Error 2
make: *** [all] Error 2
Makefile:83: recipe for target 'all' failed
Built target in bin/debug/
```

Hopson97 commented on src/common/common/obd_config.h:
```diff
@@ -0,0 +1,17 @@
+#include <unordered_map>
+
+std::unordered_map<std::string, std::string> parseObdData(const std::string& obdData);
+std::string serializeObdData(const std::unordered_map<std::string, std::string>& obdData);
```
Seeing as these will only be used by the ObdConfig internally, these function can be defined in an anonymous namespace in obd_config.cpp rather than globally available

Hopson97 commented on src/common/common/obd_config.h:
```diff
@@ -0,0 +1,17 @@
+#pragma once
+
+#include <iostream>
```
iostream not needed in this header as far as I can tell

Hopson97 commented on src/common/common/obd_config.h:
```diff
@@ -0,0 +1,17 @@
+		std::string m_filePath;
+	public:
+		ObdConfig(const std::string& filePath); // implicitly read config file
+		~ObdConfig(); // write back to file when the config goes out of scope
```
Rule of 5 should apply here, the copy constructor/assignment operator and move constructor/ assignment operator should be also defined (as =delete) 

eg

`ObdConfig(const ObdConfig&) = delete;`

etc

https://en.cppreference.com/w/cpp/language/rule_of_three

MarkusG replied:
Is this necessary if all of the resources managed by `ObdConfig` are classes?

Hopson97 commented on src/server/server_config.h:
```diff
@@ -1,8 +1,17 @@
+		void set_maxConnections(const int value) { m_maxConnections = value; }
+
+		int get_worldSize() const { return m_worldSize; }
+		void set_worldSize(const int value) { m_worldSize = value; }
```
The getters and setters are not really necessary, they can be public members/ the class can be a struct instead

Hopson97 commented on src/client/client_config.h:
```diff
@@ -1,26 +1,53 @@
+		void set_texturePack(const std::string& value) { m_texturePack = value; }
+
+		std::string get_serverIp() const { return m_serverIp; }
+		void set_serverIp(const std::string& value) { m_serverIp = value; }
```
The getters and setters are not really necessary, this class can be a struct instead with all members being public :stuck_out_tongue:

Hopson97 reviewed (changes requested):
The getters and setter are redundant really.

Few more changes

In terms of filesystem, I think that's more an issue with GitHub actions.

For now, its fine to do what you did before (Using ifstream) for checking if a file exists for the sake of the test suites passing :stuck_out_tongue: 

Thanks

MarkusG commented:
Although they might not be necessary now, the getters and setters are best practices and will be necessary in the future. For instance, the maxConnections option has a minimum value that's restricted in main, but that restriction logic could be moved to it's setter.

As for the configs being structs, can structs have constructors? I really like the idea of handling data parsing, serialization, and writing in the constructor, as well as the inheritance I've taken advantage of. Pardon my C++ ignorance, but I'm not sure which of these can still be used if we make configs structs again.

b0guslaw commented:
> Although they might not be necessary now, the getters and setters are best practices and will be necessary in the future. For instance, the maxConnections option has a minimum value that's restricted in main, but that restriction logic could be moved to it's setter.
> 
> As for the configs being structs, can structs have constructors? I really like the idea of handling data parsing, serialization, and writing in the constructor, as well as the inheritance I've taken advantage of. Pardon my C++ ignorance, but I'm not sure which of these can still be used if we make configs structs again.

Structs and classes are inherently the same concept with different default visibility. Structs are classes of which the members are all public by default, therefore a struct can have a ctor, dtor, functions and everything else a class has. I agree that getters and setters introduce an unnecessary level of verboseness

MarkusG commented:
I disagree. Given that many configuration options could have invalid data (e.g. negative screen width, negative fov, negative or 0 max players), I think it will inevitably be necessary to validate that data. Although the getters and setters do add a lot of repetitive code, they allow us to keep the validation logic in the config itself instead of having to worry about validating it everywhere we reference the config.

Hopson97 commented:
Apologies, I have only just gotten a decent chance to look at this.

After thinking about it, I agree with @MarkusG. Having setters would allow for nice validation as he says, and furthermore certain settings could be changed at run-time, and having the validation in place could prove to be incredibly useful :)

MarkusG commented:
No worries! In reference to the conversation on the rule of 5, is it necessary to define the other members if the `ObdConfig` class only manages data in the form of other C++ classes? From what I understand from https://en.cppreference.com/w/cpp/language/rule_of_three, it seems like it's only necessary if the class manages something like a raw pointer or file descriptor.

Hopson97 commented:
yes would still be needed, last thing we want is unexpected behaviour because the constructor is called on some accidentally created copy :p
